### DISCUSSION TITLE
String,String[] question when interop with tinyfiledialogs

### DISCUSSION THREAD

**Discussion by LukyGuyLucky** [c3lang/c3c#1168] [Q&A]
I attached the program package here(as I can't upload zip file in discourse) for easy reference.
At this moment,there are 2 issues I am seeking help.It is about string/char* or it is about memory management which is exactly what I would like to learn always. :)

1.  The program will crash if changing "Please input some text" to asian character 

> tinyfiledialogs::tinyfd_inputBox("query","Please input some text:","Input box");

	
2. The tinyfd_openFileDialog() will crash if not properly handling String[] conversion to char**

> 
>char* fileName=tinyfiledialogs::tinyfd_openFileDialog("let us read the password","../",2,
                 null,/*(ZString*)lFilterPatterns.ptr*///the latter will crash
>		"textfiles",1);	
>	if (!fileName)
>	{
>		              tinyfiledialogs::tinyfd_messageBox(
>			     "Error",
>			     "Open file name is NULL",
>			     "ok",
>			     "error",
>			     0);
>		      return ;
>	}
	
   

Thanks in advance,
Peter
[tinyfiledialogs_test.zip](https://github.com/c3lang/c3c/files/14517408/tinyfiledialogs_test.zip)

**Comment by lerno** [comment]
I would expect tinyfiledialogs.lib to be placed in a directory called `windows-x64` inside of the .c3l. This is usually were such files are expected to be.

These are wrong
```c
extern fn char * tinyfd_utf8toMbcs(char  * aUtf8string);
extern fn char * tinyfd_utf16toMbcs(CUInt  * aUtf16string);//CUInt
extern fn CUInt * tinyfd_mbcsTo16(char  * aMbcsString);//CUInt
extern fn char * tinyfd_mbcsTo8(char  * aMbcsString);
extern fn CUInt * tinyfd_utf8to16(char  * aUtf8string);
extern fn char * tinyfd_utf16to8(CUInt  * aUtf16string);//CUInt
```
wchar_t should be replaced by Char16, not CUInt. wchar_t can be replaced by WString and char* by ZString:

```c
extern fn char* tinyfd_utf8toMbcs(ZString aUtf8string);
extern fn char * tinyfd_utf16toMbcs(WString aUtf16string);
extern fn WString tinyfd_mbcsTo16(char* aMbcsString);
extern fn ZString tinyfd_mbcsTo8(char* aMbcsString);
extern fn WString tinyfd_utf8to16(ZString aUtf8string);
extern fn ZString tinyfd_utf16to8(WString aUtf16string);
```

`char[]` in C is basically the same as `char*`, but in C3 `char[]` is a slice.

Finally the question is what the functions expect. An Mbcs string? Strings from the source files will be UTF8.
```c
extern char tinyfd_needs[]; /* info about requirements */
```
Must be 
```c
extern char** tinyfd_needs;
```
In C3.

For this
```
//Windows only section
//#ifdef _WIN32
```
The simplest solution is to start a new module section which is only visible on Windows:

```c
//Windows only section
module tinyfiledialogs @if(env::WIN32);
```
Again all windows functions are incorrect as they use CUInt instead of Char16 (which is an ushort).

**Reply by LukyGuyLucky** [reply]
Thanks a lot.I am studying this.
